### ISSUE TITLE
Shape of model output can differ from target data shape

### ISSUE THREAD

**Issue by JohannesKnecht** [tensorflow/tensorflow#45813]
**System information**

Python 3.6, Tensorflow 2.3, Windows 10/ Ubuntu 16.04

**Describe the current behavior**

It is possible to train a model with an output shape other than the shape of the target data.
I don´t know what it´s doing there (and hope I made no mistake when thinking this trough), but shouldn´t this cause and error ?
See Section of how to reproduce.

**Describe the expected behavior**

Throw an Error if model outputshape  != target data shape.

**Standalone code to reproduce the issue**

Can be reproduced using Colab of

https://www.tensorflow.org/tutorials/keras/regression

as a basis (https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/regression.ipynb,

https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/regression.ipynb?short_path=38f0bfa)

and change Line 574
` "    layers.Dense(units=1)\n",`
to
` "    layers.Dense(units=3)\n",`

The output-shape of the Neural Network is now (batchsize, 3) however it successfully fits target data of of shape (batchsize, 1).

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/943083f25965f05ae2d8c0134562ec82/45813.ipynb). Thanks!

**Comment by around-star** [comment]
@JohannesKnecht It is owing to the fact that the loss used is "mean_squared_error".  Try using a different loss function like "categorical_crossentropy" , and you will get an error when used with different dimensions. As for the "mean_squared_error" loss function , I think it is using the (N, 1) y_true data to compare with each (N,1) in the output of (N,3) y_pred. Hope it helps.

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210602,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/4a7ec55884bc4f2a4de625d7b4351cf5/untitled183.ipynb)..Thanks !

**Comment by gowthamkpr** [comment]
@JohannesKnecht As mentioned in this [comment](https://github.com/tensorflow/tensorflow/issues/45813#issuecomment-751294380), this is an intended behavior.
